### PR TITLE
Support graph container and add functional model defination

### PR DESCRIPTION
**Pull request by yiheng**
## What changes were proposed in this pull request?
Support graph container and add functional model definition. So users can easier define more complex graph structure model.

## How was this patch tested?
add more unit tests

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+  // the outputs will be excluded
+  private val dummyOutput = new ModuleNode[T](null)
+  outputs.foreach(_ -> dummyOutput)
+  private val backGraph = dummyOutput.graph(reverse = true)
+
+  // Build execution plan
+  private val executions = backGraph.topologySort.filter(_.element != null).reverse
+  modules.appendAll(executions.map(_.element.asInstanceOf[AbstractModule[Activity, Activity, T]]))
+
+  // Make input output nodes in executions be same order with how them are passed in
```
Comment (SeaOfOcean) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala]: them -> they

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+ * will be excluded in the computation.
+ *
+ * @param inputs input nodes
+ * @param outputs output nodes
+ * @tparam T Numeric type. Only support float/double now
+ */
+class Graph[T: ClassTag](inputs : Seq[ModuleNode[T]],
+  outputs : Seq[ModuleNode[T]])(implicit ev: TensorNumeric[T])
+  extends Container[Activity, Activity, T]{
+
```
Comment (SeaOfOcean) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala]: can input same as output? (when the graph depth is 1?)

Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala]: it works. Add an unit test for this.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+ * All inputs should be able to connect to outputs through some paths in the graph. It is
+ * allowed that some successors of the inputs node are not connect to outputs. If so, these nodes
+ * will be excluded in the computation.
+ *
+ * @param inputs input nodes
+ * @param outputs output nodes
+ * @tparam T Numeric type. Only support float/double now
+ */
+class Graph[T: ClassTag](inputs : Seq[ModuleNode[T]],
+  outputs : Seq[ModuleNode[T]])(implicit ev: TensorNumeric[T])
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala]: inputs can only accept Tensor? if so, shall we check that?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+  extends Container[Activity, Activity, T]{
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala]: this is a container - what if someone calls add()?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+  }
+
+  // Add a dummy output node, to get an one end graph. So the nodes that are not dependent by
+  // the outputs will be excluded
+  private val dummyOutput = new ModuleNode[T](null)
+  outputs.foreach(_ -> dummyOutput)
+  private val backGraph = dummyOutput.graph(reverse = true)
+
+  // Build execution plan
+  private val executions = backGraph.topologySort.filter(_.element != null).reverse
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala]: no need to filer - dummy node is the first node return by topologySort(), yes?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+  modules.appendAll(executions.map(_.element.asInstanceOf[AbstractModule[Activity, Activity, T]]))
+
+  // Make input output nodes in executions be same order with how they are passed in
+  {
+    var i = 0
+    while(i < inputs.length) {
+      executions(i) = inputs(i)
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala]: what if some input is not connected to outputs?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+      i += 1
+    }
+
+    val offset = executions.length - outputs.length
+    i = 0
+    while(i < outputs.length) {
+      executions(i + offset) = outputs(i)
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala]: there is no guarantee that the last several nodes are outputs; and what if some output is not connected to inputs?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+      i += 1
+    }
+  }
+
+  private val inputsBP = new Array[Activity](executions.length)
+
+  // Check all inputs of the graph should be passed in
+  checkRoots
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala]: Instead of checkRoots, we need to check the graph more thoroughly:
1) All outputs are connected to inputs, and all inputs are connected to outputs
2) All outputs have no nextNodes, and all inputs have no prevNodes
3) There are no cycles in the graph

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+
+object Graph {
+  /**
+   * Node for graph container. The module should have a tensor/table input while a tensor output
+   * @tparam T
+   */
+  type ModuleNode[T] = Node[AbstractModule[Activity, Tensor[T], T]]
+
+  /**
+   * Build a single input, single output graph container.
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala]: wrong comment

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+   * @param output output node
+   * @return a graph container
+   */
+  def apply[T: ClassTag](input : Array[ModuleNode[T]], output : ModuleNode[T])
+    (implicit ev: TensorNumeric[T]) : Graph[T] = {
+    new Graph[T](input, Array(output))
+  }
+
+  /**
+   * Build a multiple inputs, multiple outputs graph container
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala]: wrong comment

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+    var i = 0
+    if (inputs.length == 1) {
+      // A tensor is expected here
+      inputsBP(i) = input.toTensor[T]
+      inputs(i).element.updateOutput(inputsBP(i))
+      i += 1
+    } else {
+      val inputTable = input.toTable
+      require(inputTable.length() == inputs.length,
+        "Input tensor number is not equal to inputs node number")
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala]: Graph.updateOutput:

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala [hunk 12]
```diff
@@ -0,0 +1,283 @@
...
+    var j = outputs.length - 1
+    var i = executions.length - 1
+    if (outputs.length == 1) {
+      // Expect a tensor here
+      outputs(j).element.updateGradInput(inputsBP(i), gradOutput.toTensor)
+      j -= 1
+      i -= 1
+    } else {
+      val gradOutputTable = gradOutput.toTable
+      require(gradOutputTable.length() == outputs.length,
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala]: Graph.updateGradInput:

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala [hunk 13]
```diff
@@ -0,0 +1,283 @@
...
+    var j = outputs.length - 1
+    var i = executions.length - 1
+    if (outputs.length == 1) {
+      // Expect a tensor here
+      outputs(j).element.accGradParameters(inputsBP(i), gradOutput.toTensor)
+      j -= 1
+      i -= 1
+    } else {
+      val gradOutputTable = gradOutput.toTable
+      require(gradOutputTable.length() == outputs.length,
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala]: Graph.accGradParameters:

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/DirectedGraph.scala [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+ * @param reverse use the original direction or the reversed direction
+ * @tparam T Node element type
+ */
+class DirectedGraph[T](source : Node[T], reverse : Boolean = false) {
+
+  /**
+   * Topology sort.
+   * @return A sequence of sorted graph nodes
+   */
+  def topologySort : Array[Node[T]] = {
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/DirectedGraph.scala]: topologySort  can be implemented using DFS

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala (hunk 3 above)
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala]: If there are multiple inputs, the input data fed to the graph module should be a table, which is actually an sequence of tensor. The order of the input tensors should be same with the order of the input nodes. 

We have code check that. Have added two unit test to cover this.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala (hunk 4 above)
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala]: Good questions... Add makes no sense in the graph container. I can override it and throw exceptions to hint user.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala (hunk 11 above)
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala]: changed. But I don't quite understand this as you can see where is the error from in the stack, why need to add Graph.updateOutput?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala (hunk 5 above)
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala]: Yes. But it's easy to write some following code if we exclude it. e.g. if you don't filter here, you have to filter in the next line: modules.appendAll. It's one time work, don't need to worry the perf

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala (hunk 6 above)
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala]: We needn't guarantee these nodes are first of the total graph. User can specify a part of graph to Graph container for computing. It's common use in tf and keras.

> If some input is not connected to outpus, there' will be error in graph init. And we have covered this in the doc and unit test.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala (hunk 7 above)
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala]: We needn't guarantee these nodes are first of the total graph. User can specify a part of graph to Graph container for computing. It's common use in tf and keras.

> If some input is not connected to outpus, there' will be error in graph init. And we have covered this in the doc and unit test.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala (hunk 8 above)
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala]: for 1,
>  All inputs should be able to connect to outputs through some paths in the graph. It is allowed that some successors of the inputs node are not connect to outputs. If so, these nodes will be excluded in the computation.

for 2,
I don't think we need do such limitation. User can specify a part of graph to Graph container for computing. It's common use in tf and keras.

for 3
It is already check in the topology sort.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/DirectedGraph.scala (hunk 14 above)
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/DirectedGraph.scala]: I check the output of the DFS based topology sort. Notice that it may loss the parallel property. If we use the indegree based topology sort, we can know which nodes can be run in parallel. This property is useless now as we run a single model per thread but I think it's a useful property to have.

This remind me I should build a separated execution plan for forward and backward.
